snmalloc-edp/src/rust-sgx-snmalloc-shim.cpp

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) Microsoft Corporation.
+// Copyright (c) Open Enclave SDK contributors.
+// Copyright (c) 2020 SchrodingerZhu
+// Copyright (c) Fortanix, Inc.
+//
+// MIT License
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE
+
+#include <immintrin.h>
+#include <string.h>
+
+/***************************************************/
+/*** Imported symbols needed by snmalloc SGX PAL ***/
+/***************************************************/
+
+// from entry.S
+extern "C" size_t get_tcs_addr();
+
+// from Rust std
+extern "C" void __rust_print_err(const char *m, size_t s);
+extern "C" [[noreturn]] void __rust_abort();
+
+/*******************************************************/
+/*** Standard C functions needed by snmalloc SGX PAL ***/
+/*******************************************************/
+
+// definition needs to match GNU header
+extern "C" [[noreturn]] void abort() __THROW {
+    __rust_abort();
+}
+
+// definition needs to match GNU header
+extern "C" inline int * __attribute_const__ __errno_location (void) __THROW {
+    static int errno;
+    return &errno;
+}
+
+/***********************************/
+/*** snmalloc SGX PAL definition ***/
+/***********************************/
+
+#define SNMALLOC_PROVIDE_OWN_CONFIG
+#define SNMALLOC_SGX
+#define SNMALLOC_USE_SMALL_CHUNKS
+#define SNMALLOC_MEMORY_PROVIDER PALEdpSgx
+
+#include "../snmalloc/src/snmalloc/pal/pal_noalloc.h"
+
+namespace snmalloc {
+void register_clean_up() {
+    // TODO: not sure what this is supposed to do
```

Comment:
We probably need to resolve this before upstreamining.

Reply:
I'm not 100% sure on this, but the only place I see `register_clean_up()` being called is if the allocator supports lazy initialization. I'm assuming our allocator doesn't support lazy initialization, so I think we probably don't need our `register_clean_up()` to do anything.